modules/gdscript/gdscript_analyzer.cpp

```diff
@@ -1601,6 +1481,132 @@ void GDScriptAnalyzer::resolve_suite(GDScriptParser::SuiteNode *p_suite) {
+			} else if (initializer_type.kind == GDScriptParser::DataType::BUILTIN && initializer_type.builtin_type == Variant::NIL && !is_constant) {
+				push_error(vformat(R"(Cannot infer the type of "%s" %s because the value is "null".)", p_assignable->identifier->name, p_kind), p_assignable->initializer);
+			}
+		} else {
+			if (!initializer_type.is_set()) {
+				push_error(vformat(R"(Could not resolve type for %s "%s".)", p_kind, p_assignable->identifier->name), p_assignable->initializer);
+			}
+		}
+
+		if (!has_specified_type) {
+			type = initializer_type;
+
+			if (!type.is_set() || (type.is_hard_type() && type.kind == GDScriptParser::DataType::BUILTIN && type.builtin_type == Variant::NIL && !is_constant)) {
+				type.kind = GDScriptParser::DataType::VARIANT;
+			}
+
+			if (p_assignable->infer_datatype || is_constant) {
+				type.type_source = GDScriptParser::DataType::ANNOTATED_INFERRED;
+			} else {
+				type.type_source = GDScriptParser::DataType::INFERRED;
+			}
+		} else if (!specified_type.is_variant()) {
+			if (initializer_type.is_variant() || !initializer_type.is_hard_type()) {
+				mark_node_unsafe(p_assignable->initializer);
+				if (is_variable) {
+					static_cast<GDScriptParser::VariableNode *>(p_assignable)->use_conversion_assign = true;
+				}
+			} else if (!is_type_compatible(specified_type, initializer_type, true, p_assignable->initializer)) {
+				if (is_variable && is_type_compatible(initializer_type, specified_type, true, p_assignable->initializer)) {
+					mark_node_unsafe(p_assignable->initializer);
+					static_cast<GDScriptParser::VariableNode *>(p_assignable)->use_conversion_assign = true;
+				} else {
+					push_error(vformat(R"(Cannot assign a value of type %s to %s "%s" with specified type %s.)", initializer_type.to_string(), p_kind, p_assignable->identifier->name, specified_type.to_string()), p_assignable->initializer);
+				}
+#ifdef DEBUG_ENABLED
+			} else if (specified_type.builtin_type == Variant::INT && initializer_type.builtin_type == Variant::FLOAT) {
+				parser->push_warning(p_assignable->initializer, GDScriptWarning::NARROWING_CONVERSION);
+#endif
+			}
+		}
+	}
+
+	type.is_constant = is_constant;
+	p_assignable->set_datatype(type);
+}
+
+void GDScriptAnalyzer::resolve_variable(GDScriptParser::VariableNode *p_variable, bool p_is_local) {
+	static constexpr const char *kind = "variable";
+	resolve_assignable(p_variable, kind);
+
+#ifdef DEBUG_ENABLED
+	if (p_is_local) {
+		if (p_variable->usages == 0 && !String(p_variable->identifier->name).begins_with("_")) {
+			parser->push_warning(p_variable, GDScriptWarning::UNUSED_VARIABLE, p_variable->identifier->name);
+		} else if (p_variable->assignments == 0) {
+			parser->push_warning(p_variable, GDScriptWarning::UNASSIGNED_VARIABLE, p_variable->identifier->name);
+		}
+
+		is_shadowing(p_variable->identifier, "variable");
```

Comment:
You could use `kind` here instead of repeating the string `"variable"`. Same for `"constant"` below.

On an unrelated note, we should maybe rename the `is_shadowing()` function given it has side effects and we never use its result.

Reply:
Done. (Renaming note sounds reasonable.)